engine/utils/HeapArray.h

```diff
@@ -0,0 +1,735 @@
+             * For example, given a bit position in index 1, we create a bit mask with a value of
+             * 1 left shifted by 1 position (0000 0010). Finally, since our bytes contain 8 bits,
+             * we want to make sure that we just check the index we get is relative to our byte.
+             * i.e: 2 % 8 = 2 or 9 % 8 = 1. No matter how large the index, we'll always get its
+             * position in 8s.
+             */
+            auto bitPosition = (1 << ((index + 1) % BYTE_SIZE_IN_BITS));
+
+            /**
+             * Run a bitwise AND operation to check if the corresponding byte exists in the original
+             * mask. i.e: 1010 0010 & 0000 0010 will return 1 (since 1 exists in the second bit).
+             *
+             * Finally, we also want to make sure the next index is in bounds.
+             */
+            return (byteMask & bitPosition) == 0 && (index + 1) < size;
+        }
+
+        /**
+         * @brief An equal operator, checks with equality.
+         * @param other the other iterator to compare to.
+         * @return `true` if the pointers stored by both iterators are the same.
+         */
+        bool operator==(const Iterator& other) const
+        {
+            return pointer == other.pointer;
+        }
+
+        /**
+         * @brief A non equality operator. Checks that twio iterators are not equal.
+         * @param other the iterator to compare to.
+         * @return `true` if the pointer addresses stored by both iterators are not the same.
+         */
+        bool operator!=(const Iterator& other) const
+        {
+            return pointer != other.pointer;
+        }
+
+    private:
+
+        T* pointer;
+        uint8_t* maskPointer;
+        size_t index;
+        size_t size;
+    };
+
+    /**
+     * @brief An empty constructor. This is invoked when no parameters are passed to the
+     * constructor. Initialises the array with a size of zero. Does not allocate any space on the
+     * heap. Constructing an array this way requires either an explicit resizing or a new creation
+     * to add a size. Any usage of the Get() or subscript operator will cause an error.
+     */
+    HeapArray() : HeapArray(0, 0, 0, nullptr, nullptr) {}
+
+    /**
+     * @brief Initialises the HeapArray with the given size.
+     * @param arraySize the size of the array.
+     */
+    explicit HeapArray(const size_t& arraySize) :
+        HeapArray(arraySize, 0, CalculateByteMaskCount(arraySize))
+    {}
+
+    /**
+     * @brief Initialises the array using an initialiser list.
+     * @param list the initialiser list.
+     */
+    HeapArray(const std::initializer_list<T>& list)
+    {
+        size = list.size();
+
+        data = Allocate<T>(sizeof(T) * list.size());
+
+        size_t newByteCount = CalculateByteMaskCount(list.size());
+
+        stateMasks = Allocate<uint8_t>(BYTE_MASK_SIZE * newByteCount);
+
+        memset(stateMasks, 0, (BYTE_MASK_SIZE * newByteCount));
```

Comment:
This pattern of allocation and memset would not be an issue if it weren't a pointer.

Reply:
Again, it has to be a pointer because we're storing an array of bitmasks, not a single bitmask.

Reply:
Then roll these two into a function if it's going to be done multiple times across the class.

Reply:
sure thang - I'll get to this over the weekend since I've run out of time.

Reply:
Just had a look - this pattern is not repeated anywhere else in the class.